File: src/x/crazyjvm/strategy/j/Context.java
package x.crazyjvm.strategy.j;

/**
 * Created by chenchao on 15/10/13.
 */
public class Context{
    private final Strategy strategy;

    public Context(Strategy strategy) { this.strategy = strategy; }

    public void use(int a, int b) { strategy.compute(a, b); }

    public static void main(String[] args){
        new Context(new Multiply()).use(2, 3);
    }
}

File: src/x/crazyjvm/strategy/s/Context.scala
package x.crazyjvm.strategy.s

/**
 * Created by chenchao on 15/10/13.
 */
object Main extends App{
  type Strategy = (Int, Int) => Int

  class Context(computer: Strategy) {
    def use(a: Int, b: Int)  { computer(a, b) }
  }

  val add: Strategy = _ + _
  val multiply: Strategy = _ * _

  new Context(multiply).use(2, 3)
}
